command/agent/exec/exec.go

```diff
@@ -0,0 +1,260 @@
+				s.proc.Stop()
+			}
+			s.procState = processStateStopped
+			return nil
+		case token := <-incomingVaultToken:
+			if token != *latestToken {
+				s.logger.Info("exec server received new token")
+
+				s.runner.Stop()
+				*latestToken = token
+				newTokenConfig := ctconfig.Config{
+					Vault: &ctconfig.VaultConfig{
+						Token:           latestToken,
+						ClientUserAgent: pointerutil.StringPtr(useragent.AgentTemplatingString()),
+					},
+				}
+
+				// got a new auth token, merge it in with the existing config
+				runnerConfig = runnerConfig.Merge(&newTokenConfig)
+				s.runner, err = manager.NewRunner(runnerConfig, true)
+				if err != nil {
+					s.logger.Error("template server failed with new Vault token", "error", err)
+					continue
+				}
+				go s.runner.Start()
+			}
+
+		case err := <-s.runner.ErrCh:
+			s.logger.Error("template server error", "error", err.Error())
+			s.runner.StopImmediately()
+
+			// Return after stopping the runner if exit on retry failure was specified
+			if s.config.AgentConfig.TemplateConfig != nil && s.config.AgentConfig.TemplateConfig.ExitOnRetryFailure {
+				return fmt.Errorf("template server: %w", err)
+			}
+
+			s.runner, err = manager.NewRunner(runnerConfig, true)
+			if err != nil {
+				return fmt.Errorf("template server failed to create: %w", err)
+			}
+			go s.runner.Start()
+		case <-s.runner.TemplateRenderedCh():
+			// A template has been rendered, figure out what to do
+			s.logger.Debug("template rendered")
+			events := s.runner.RenderEvents()
+
+			// events are keyed by template ID, and can be matched up to the id's from
+			// the lookupMap
+			if len(events) < s.numberOfTemplates {
+				// Not all templates have been rendered yet
+				continue
+			}
+
+			// assume the renders are finished, until we find otherwise
+			doneRendering := true
+			var renderedEnvVars []string
+			for _, event := range events {
+				// This template hasn't been rendered
+				if event.LastWouldRender.IsZero() {
+					doneRendering = false
+					break
+				} else {
+					for _, tcfg := range event.TemplateConfigs {
+						envVar := fmt.Sprintf("%s=%s", *tcfg.MapToEnvironmentVariable, event.Contents)
+						renderedEnvVars = append(renderedEnvVars, envVar)
+					}
+				}
+			}
+
+			if doneRendering {
+				s.logger.Debug("done rendering templates/detected change, bouncing process")
+				if err := s.bounceCmd(renderedEnvVars); err != nil {
+					return fmt.Errorf("unable to bounce command: %w", err)
+				}
+			}
+		case exitCode := <-s.procExitCh:
+			// process exited on its own
+			return &ProcessExitError{ExitCode: exitCode}
+		}
+	}
+}
+
+func (s *Server) bounceCmd(newEnvVars []string) error {
+
+	switch s.config.AgentConfig.Exec.RestartOnSecretChanges {
+	case "always":
+		if s.procState == processStateRunning {
+			// process is running, need to kill it first
+			s.logger.Info("stopping process", "process_id", s.proc.Pid())
+			s.proc.Stop()
+			s.procState = processStateRestarting
```

Comment:
```suggestion
			s.procState = processStateRestarting
			s.proc.Stop()
```

May want to swap the two lines in case `Stop` is a blocking call.